docs/docs/reference-react-component.md

```diff
@@ -230,34 +230,59 @@ componentWillUnmount()
 ### `setState()`
 
 ```javascript
-setState(nextState, callback)
+setState(updater, [callback])
 ```
 
-Performs a shallow merge of nextState into current state. This is the primary method you use to trigger UI updates from event handlers and server request callbacks.
+`setState()` is an asynchronous method which enqueues changes to be made to the state during the next update cycle. This is the primary method you use to trigger updates in response to event handlers, server responses, etc...
 
-The first argument can be an object (containing zero or more keys to update) or a function (of state and props) that returns an object containing keys to update.
+`setState()` does not immediately mutate `this.state` but creates a pending state transition. Accessing `this.state` after calling this method can potentially return the previous state, rather than the state after enqueued updates have been applied. This is a common source of bugs in React applications.
```

Comment:
I know this from mentoring students & product teams who are making the switch to React. I saw two examples yesterday. See [setState Gate](https://medium.com/javascript-scene/setstate-gate-abc10a9b2d82). Further evidence is available by searching for [setState](http://stackoverflow.com/questions/28922275/in-reactjs-why-does-setstate-behave-differently-when-called-synchronously) [on](http://stackoverflow.com/questions/34820583/is-reacts-setstate-asynchronous-or-something) [Stack Overflow](http://stackoverflow.com/questions/41896878/react-setstate-getstate-and-asynchronous) (there are MANY more examples of confused setState users on Stack Overflow).